WalletWasabi/Blockchain/Transactions/SmartTransaction.cs

```diff
@@ -83,6 +92,54 @@ public IReadOnlyCollection<IndexedTxOut> ForeignOutputs
 		}
 	}
 
+	/// <summary>Wallet inputs with the same script are virtually considered to be the same by blockchain analysis.</summary>
+	public IReadOnlyCollection<WalletVirtualInput> WalletVirtualInputs
+	{
+		get
+		{
+			if (WalletVirtualInputsCache is null)
+			{
+				WalletVirtualInputsCache = WalletInputs
+					.GroupBy(i => i.HdPubKey.PubKeyHash.ToBytes())
+					.Select(g => new WalletVirtualInput(g.Key, g.ToHashSet()))
+					.ToHashSet();
+			}
+			return WalletVirtualInputsCache;
+		}
+	}
+
+	/// <summary>Wallet outputs with the same script are virtually considered to be the same by blockchain analysis.</summary>
+	public IReadOnlyCollection<WalletVirtualOutput> WalletVirtualOutputs
+	{
+		get
+		{
+			if (WalletVirtualOutputsCache is null)
+			{
+				WalletVirtualOutputsCache = WalletOutputs
+					.GroupBy(o => o.HdPubKey.PubKeyHash.ToBytes())
+					.Select(g => new WalletVirtualOutput(g.Key, g.Sum(o => o.Amount), g.Select(o => new OutPoint(GetHash(), o.Index)).ToHashSet()))
```

Comment:
I'm not saying it is a big deal but you are calculating the same transaction hash for each transaction output separately which leads to quadratic complexity.

Reply:
It's ok, `GetHash()` does not calculate anything, it's already there, the only reason why it's a function and not a property is because I wanted to be consistent with NBitcoin's `Transaction` object, which is being created with "precompute hash" parameter, so we're fine.